### PR TITLE
Building ubuntu and debian inside docker

### PR DESCRIPTION
**Pull request by shimwell**
Following on from the last DAGMC meeting here is a PR that attempts to build the plugin for debian 10.10

The approach taken is to use a docker container to build linux flavours (ubuntu 21.04, ubuntu 20.04, debian 10.10) and to use the base github actions (without a docker container) to build for mac os.

This appears to work given the constraints:
- containerization of mac is not currently possible

This appears to help with #112 as the plugin can be build for a few more OS and OS versions.

### CODE REVIEW

**Comment by shimwell** [comment]
Everything appears to be building 🎉 thanks @bam241 for getting this started

File: .github/workflows/unix_linux.yml [hunk 1]
```diff
@@ -18,23 +18,49 @@ env:
...
+          if [ "${{ matrix.os }}" == "ubuntu" ]; then
+            apt-get update
+            apt-get install -y software-properties-common
+            add-apt-repository ppa:git-core/ppa
+            apt-get update
+            apt-get install -y git
+          elif [ "${{ matrix.os }}" == "debian" ]; then
+            apt update
+            apt-get install -y software-properties-common
+            apt update
```
Comment (gonuke) [inline .github/workflows/unix_linux.yml]: Is this update necessary?

File: .github/workflows/unix_linux.yml [hunk 2]
```diff
@@ -104,6 +120,7 @@ jobs:
         run: |
           mkdir -p $FOLDER_PKG
           cd ${FOLDER_PKG}
+          apt-get install -y curl
```
Comment (gonuke) [inline .github/workflows/unix_linux.yml]: Can we add this to the earlier apt-get?

File: .github/workflows/unix_linux.yml [hunk 3]
```diff
@@ -57,26 +83,16 @@ jobs:
             HDF5_PATH=/usr/local/HDF_Group/HDF5/1.12.0
           fi
 
-          if [[ "${{ matrix.os }}" == "ubuntu"*  ]]; then
+          if [ "${{ matrix.os }}" == "debian" ] || [ "${{ matrix.os }}" == "ubuntu" ]; then
             SUFFIX=Lin64
             EXT=deb
             echo "SED=sed" >> $GITHUB_ENV
-            echo "SUDO=sudo " >> $GITHUB_ENV
```
Comment (gonuke) [inline .github/workflows/unix_linux.yml:64]: Why don't we want SUDO defined?

File: .github/workflows/unix_linux.yml [hunk 4]
```diff
@@ -57,26 +83,16 @@ jobs:
             HDF5_PATH=/usr/local/HDF_Group/HDF5/1.12.0
           fi
 
-          if [[ "${{ matrix.os }}" == "ubuntu"*  ]]; then
+          if [ "${{ matrix.os }}" == "debian" ] || [ "${{ matrix.os }}" == "ubuntu" ]; then
```
Comment (gonuke) [inline .github/workflows/unix_linux.yml]: Do we need a conditional still if we remove Mac? Or we add some failure path if we don't match this condition

**Review by gonuke — CHANGES_REQUESTED** [review]
A few changes to clean things up.

At some point we should review the overall design to increase reuse of things rather than maintaining different scripts.

File: .github/workflows/unix_linux.yml (hunk 3 above)
Comment (shimwell) [reply .github/workflows/unix_linux.yml:64]: I'm not actually sure about this part, it was in the original script.

Comment (gonuke) [reply .github/workflows/unix_linux.yml:64]: We probably still want it

Comment (shimwell) [reply .github/workflows/unix_linux.yml:64]: ok, we can leave this in for now, possibly come back to it one day

File: .github/workflows/unix_linux.yml (hunk 2 above)
Comment (shimwell) [reply .github/workflows/unix_linux.yml]: now added to earlier apt-get

File: .github/workflows/unix_linux.yml (hunk 4 above)
Comment (shimwell) [reply .github/workflows/unix_linux.yml]: now removed this condition. Perhaps we will bring this back one day it more OS get added

File: .github/workflows/unix_linux.yml (hunk 1 above)
Comment (shimwell) [reply .github/workflows/unix_linux.yml]: reworked this section to reduce number of updates

**Comment by shimwell** [comment]
I've addressed all the suggestions and done as recommended, thanks for the review. Let's see how it does with the CI building process.

**Comment by shimwell** [comment]
> A few changes to clean things up.
> 
> At some point we should review the overall design to increase reuse of things rather than maintaining different scripts.

Sounds like a good plan, I think once GH actions allows containers of Mac OS there is some recombination that can be done

**Comment by bam241** [comment]
There some issues with the plug-in artifacts uploads: https://github.com/svalinn/Cubit-plugin/actions/runs/1114012215
And:

https://github.com/svalinn/Cubit-plugin/actions/runs/1114012218

One also can test those individually to ensure they are working as intended :)

**Comment by bam241** [comment]
I meant test them locally

**Comment by bam241** [comment]
> > A few changes to clean things up.
> 
> > 
> 
> > At some point we should review the overall design to increase reuse of things rather than maintaining different scripts.
> 
> 
> 
> Sounds like a good plan, I think once GH actions allows containers of Mac OS there is some recombination that can be done

This is very unlikely to happen as it will broken general usage conditions of MacOS: you can only use macOS on a Mac machine... 
I.e. virtualization of macOS is only allowed on a Mac host.... allowing it on docker will imply requiring a lot of safeguards on the docker implementation side to make sure you cannot run a macOS is on a non-Mac host...
Multi-platform virtualizer apps tended to stay away from those complication in the past

File: .github/workflows/unix_linux.yml [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+      - name: Prepare package
+        shell: bash -l {0}
+        run: |
+          ${system}_build_plugin_pkg ${{ matrix.cubit }}
+
+      - if: github.event_name != 'release'
+        name: Upload artifact for CI
+        uses: actions/upload-artifact@v2
+        with:
+          name: svalinn-plugin_${{ matrix.os }}-${{ matrix.os_version }}_cubit_${{ matrix.cubit }}.tgz
```
Comment (bam241) [inline .github/workflows/unix_linux.yml:152]: those artifact names needs to mach the one produced by the `${system}_build_plugin_pkg` macro.
I think the macro need to be tweaked to accept `os` and `os_version` variables.

File: .github/workflows/unix_linux.yml [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+          path: ${{ github.workspace }}/release/svalinn-plugin_${{ matrix.os }}-${{ matrix.os_version }}_cubit_${{ matrix.cubit }}.tgz
```
Comment (bam241) [inline .github/workflows/unix_linux.yml:153]:
```suggestion
          path: ${{ github.workspace }}/release/svalinn-plugin_${{ matrix.os }}-${{ matrix.os_version }}_cubit_${{ matrix.cubit }}.tgz
          if-no-files-found: error # 'warn' or 'ignore' are also available, defaults to `warn` 
```

Comment (bam241) [reply .github/workflows/unix_linux.yml:153]: this will allow to be aware when we have a problem founding the tgz file

Comment (shimwell) [reply .github/workflows/unix_linux.yml:153]: Thanks, this is a nice check.

File: .github/workflows/unix_linux.yml (hunk 5 above)
Comment (shimwell) [reply .github/workflows/unix_linux.yml:152]: Thanks, I see the ```unix_share_build.sh``` current finds the linux (major) version using ```LINUX_VERSION=$(lsb_release -rs |cut -d"." -f1)``` . Should I remove the ```lsb_release``` part and pass in ```os_version``` as predefined environmental variable

File: .github/workflows/unix_mac.yml [hunk 7]
```diff
@@ -144,14 +126,14 @@ jobs:
         name: Upload artifact for CI
         uses: actions/upload-artifact@v2
         with:
-          name: svalinn-plugin_${{ matrix.os }}_cubit_${{ matrix.cubit }}.tgz
-          path: ${{ github.workspace }}/release/svalinn-plugin_${{ matrix.os }}_cubit_${{ matrix.cubit }}.tgz
+          name: svalinn-plugin_${{ matrix.os }}-${{ matrix.os_version }}_cubit_${{ matrix.cubit }}.tgz
+          path: ${{ github.workspace }}/release/svalinn-plugin_${{ matrix.os }}-${{ matrix.os_version }}_cubit_${{ matrix.cubit }}.tgz
```
Comment (bam241) [inline .github/workflows/unix_mac.yml:130]:
```suggestion
          path: ${{ github.workspace }}/release/svalinn-plugin_${{ matrix.os }}-${{ matrix.os_version }}_cubit_${{ matrix.cubit }}.tgz
          if-no-files-found: error # 'warn' or 'ignore' are also available, defaults to `warn` 
```

**Review by bam241 — CHANGES_REQUESTED** [review]
this change would be required for Mac as well :)

File: scripts/unix_share_build.sh [hunk 8]
```diff
@@ -25,19 +24,26 @@ function mac_install_prerequisites() {
...
 }
 
 function linux_install_prerequisites() {
     TZ=America/Chicago
     $SUDO ln -snf /usr/share/zoneinfo/$TZ /etc/localtime
     $SUDO sh -c 'echo $TZ > /etc/timezone'
     $SUDO apt-get update -y
-    $SUDO apt-get install -y g++ libeigen3-dev patchelf git cmake curl lsb-release python3  lsb-core
+    if [ "$OS" == "ubuntu" ]; then
+    $SUDO apt-get install -y g++ libeigen3-dev patchelf git cmake curl lsb-release python3 lsb-core
```
Comment (gonuke) [inline scripts/unix_share_build.sh:39]: Notice that this script relies on `SUDO` but it never gets set.  Current behavior is probably that this resolves to nothing and doesn't matter... but we should either remove the `$SUDO` here, or set the variable as appropriate.

**Review by gonuke — CHANGES_REQUESTED** [review]
Something happened to SUDO

File: .github/workflows/unix_linux.yml (hunk 5 above)
Comment (shimwell) [reply .github/workflows/unix_linux.yml:152]: Went for the option of passing in another variable called os_version in #116

File: scripts/unix_share_build.sh (hunk 8 above)
Comment (shimwell) [reply scripts/unix_share_build.sh:39]: I've now set SUDO to an empty string in the mac and linux github actions scripts in commit 6dc6a66

Comment (shimwell) [reply scripts/unix_share_build.sh:39]: Ah that setting of SUDO to an empty string appears to have not worked

**Comment by shimwell** [comment]
Looks like Trelis 17.1.0 is failing on Ubuntu 21.04 due to not finding HDF5

```Could NOT find HDF5 (missing: HDF5_LIBRARIES HDF5_INCLUDE_DIRS) (found  version "")```

This was not actually build previously so I shall add it to the exclude list.

Perhaps it is fair to exclude combinations of old cubit with new os or vice versa

Ubuntu 20.04 with Cubit 17.0.1 was previously excluded, so I think it is reasonable to also exclude 21.04 with this version of Cubit

**Comment by shimwell** [comment]
Builds are passing for all the combinations 🎉

File: .github/workflows/unix_linux.yml [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+  GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+jobs:
+  building-plugin:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        cubit: [2021.4, 2021.5]
+        os: [ubuntu]
+        os_version: [20.04, 21.04]
```
Comment (gonuke) [inline .github/workflows/unix_linux.yml:27]: Can we add in 18.04 here?

**Review by gonuke — CHANGES_REQUESTED** [review]
Just wondering about another Ubuntu version for building?

File: .github/workflows/unix_linux.yml (hunk 9 above)
Comment (shimwell) [reply .github/workflows/unix_linux.yml:27]: Certainly

Comment (shimwell) [reply .github/workflows/unix_linux.yml:27]: Ah looks like 18.04 is failing with ```Could NOT find HDF5 (missing: HDF5_LIBRARIES HDF5_INCLUDE_DIRS) (found  version "")```

Comment (gonuke) [reply .github/workflows/unix_linux.yml:27]: Looking at the currently working version... it appears that we didn't support 18.04 with anything other than Cubit 17.  @bam241 will remember why...  You can remove it again.  I think this new version (without 18.04 in the matrix) will build at least as many combinations as the old version.

Comment (shimwell) [reply .github/workflows/unix_linux.yml:27]: Yep I think this new version builds everything the old version did and also a 4 extras
Ubuntu 21.04 Cubit 2021.4
Ubuntu 21.04 Cubit 2021.5
Debian 10.10 Cubit 2021.5
Mac OS 10.15 Cubit 2021.5

Comment (shimwell) [reply .github/workflows/unix_linux.yml:27]: |              | Trelis 17.01       | Cubit 2021.4 | Cubit 2021.5 |
|--------------|--------------------|--------------|--------------|
| Ubuntu 18.04 | :heavy_check_mark: |              |              |
| Ubuntu 20.04 |                    |  :heavy_check_mark:|  :heavy_check_mark:|
| Ubuntu 21.04 |                    |  :heavy_check_mark:|  :heavy_check_mark:|
| Debian 10.10 |                    |              |  :heavy_check_mark:|
| Mac OS 10.15 |  :heavy_check_mark:|  :heavy_check_mark:|  :heavy_check_mark:|

**Comment by gonuke** [comment]
Thanks @Shimwell - this gives us a good starting point to refine and grow

**Pull request merged**